### ISSUE TITLE
Tests depend on specific `TZ` setting

### ISSUE THREAD

**Issue by ccorn** [kellyjonbrazil/jc#148]
Try

    TZ=UTC python setup.py test

This should cause timestamp-parsing-related test failures.

I suppose that the timestamp parser relies on the current `TZ` environment setting being the same as the one used for generating the timestamp, which might not be the case for the pre-recorded test input. I had to explicitly run

    TZ=America/Los_Angeles python setup.py test

to make the tests succeed.

IMHO, the test script should take care of setting the expected timezone.

**Comment by kellyjonbrazil** [comment]
Hi there,

Yes, this is a known issue and I've been thinking of how to address this in a cross-platform way (linux/macOS vs. Windows). Today I set the timezone locally or in github actions before the test, but I've been thinking about how to accurately detect the timezone and ignore certain tests or perhaps to set the timezone from within the test scripts.

**Comment by kellyjonbrazil** [comment]
Maybe I can do something like this in the tests that require the specific timezone:
```
if sys.platform.startswith('win32'):
    os.system('tzutil /s "Pacific Standard Time"')
else:
    os.environ['TZ'] = 'America/Los_Angeles'
    time.tzset()
```

**Comment by ccorn** [comment]
> Maybe I can do something like this in the tests that require the specific timezone:
> 
> ```
> if sys.platform.startswith('win32'):
>     os.system('tzutil /s "Pacific Standard Time"')
> ```

If `tzutil /s` has effects beyond the test, I'd rather not use that.

**Comment by kellyjonbrazil** [comment]
Good point, though I'm not sure what the options in Windows would be other than to ignore the tests if the timezone is not already correct. I could store the existing timezone using `tzutil /g` and set it back after the test, but that would temporarily set it for the whole system, which may not be desirable.

**Comment by ccorn** [comment]
Part of your proposal should help on POSIX-like systems:

```python
if not sys.platform.startswith('win32'):
    os.environ['TZ'] = 'America/Los_Angeles'
    time.tzset()
```

No idea how to handle Windows, as I am not aware of a per-process way.

Alternatively, if python's date parser itself could be explicitly configured with a timezone spec, that would help on all platforms.

**Comment by kellyjonbrazil** [comment]
Cool - I used your code in the impacted test files. POSIX should be taken care of, but I decided to punt on Windows for now. Updates are in the `dev` branch. Thanks!

**Comment by ccorn** [comment]
Tested the updated `dev` branch (#3e54b597be88b0e342e4361b7f2c1af447f9482b) on `x86_64-linux`, and the tests succeed without me having to set `TZ` explicitly. Thanks!
For me, this is sufficient. The changes are not on `master` yet and perhaps a side-effect-free way to deal with Windows can be found, so I do not close this issue now. Instead I leave it to you to close this issue when that seems appropriate.

**Comment by kellyjonbrazil** [comment]
Thanks - I'll close this for now. If this is enough of an issue on Windows I'm sure someone will open another issue. As you mention, it might be tough to find a way to do it without clobbering the timezone on Windows for the entire system.

**Comment by chocolateboy** [comment]
FYI, this is still an issue on Linux, with the [AUR package](https://aur.archlinux.org/packages/jc) at least, which hasn't successfully installed for me since 1.14.0 (January 2021):

```
======================================================================
FAIL: test_date (tests.test_date.MyTests)
Test 'date'
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/jc/src/jc-1.18.2/tests/test_date.py", line 61, in test_date
    self.assertEqual(jc.parsers.date.parse(self.generic_date, quiet=True), self.generic_date_json)
AssertionError: {'yea[147 chars]d': 'am', 'timezone': 'PDT', 'utc_offset': Non[132 chars]alse} != {'yea[147 chars]d': 'AM', 'timezone': 'PDT', 'utc_offset': Non[132 chars]alse}
  {'day': 3,
   'day_of_year': 216,
   'epoch': 1596471171,
   'epoch_utc': None,
   'hour': 9,
   'hour_24': 9,
   'iso': '2020-08-03T09:12:51',
   'minute': 12,
   'month': 'Aug',
   'month_num': 8,
-  'period': 'am',
?             ^^

+  'period': 'AM',
?             ^^

   'second': 51,
   'timezone': 'PDT',
   'timezone_aware': False,
   'utc_offset': None,
   'week_of_year': 31,
   'weekday': 'Mon',
   'weekday_num': 1,
   'year': 2020}

======================================================================
FAIL: test_date_after_midnight (tests.test_date.MyTests)
Test 'date' 24-hour conversion just after midnight
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/jc/src/jc-1.18.2/tests/test_date.py", line 73, in test_date_after_midnight
    self.assertEqual(jc.parsers.date.parse(self.generic_date_after_midnight, quiet=True), self.generic_date_after_midnight_json)
AssertionError: {'yea[149 chars]d': 'am', 'timezone': 'UTC', 'utc_offset': '+0[145 chars]True} != {'yea[149 chars]d': 'AM', 'timezone': 'UTC', 'utc_offset': '+0[145 chars]True}
  {'day': 24,
   'day_of_year': 83,
   'epoch': 1616572487,
   'epoch_utc': 1616547287,
   'hour': 12,
   'hour_24': 0,
   'iso': '2021-03-24T00:54:47+00:00',
   'minute': 54,
   'month': 'Mar',
   'month_num': 3,
-  'period': 'am',
?             ^^

+  'period': 'AM',
?             ^^

   'second': 47,
   'timezone': 'UTC',
   'timezone_aware': True,
   'utc_offset': '+0000',
   'week_of_year': 12,
   'weekday': 'Wed',
   'weekday_num': 3,
   'year': 2021}

======================================================================
FAIL: test_date_am_ubuntu_20_04 (tests.test_date.MyTests)
Test 'date' on Ubuntu 20.4 with LANG=en_US.UTF-8 (uses 12-hour clock) with AM time
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/jc/src/jc-1.18.2/tests/test_date.py", line 79, in test_date_am_ubuntu_20_04
    self.assertEqual(jc.parsers.date.parse(self.ubuntu_20_04_date, quiet=True), self.ubuntu_20_04_date_json)
AssertionError: {'yea[145 chars]d': 'am', 'timezone': 'UTC', 'utc_offset': '+0[143 chars]True} != {'yea[145 chars]d': 'AM', 'timezone': 'UTC', 'utc_offset': '+0[143 chars]True}
  {'day': 5,
   'day_of_year': 5,
   'epoch': 1609837324,
   'epoch_utc': 1609808524,
   'hour': 1,
   'hour_24': 1,
   'iso': '2021-01-05T01:02:04+00:00',
   'minute': 2,
   'month': 'Jan',
   'month_num': 1,
-  'period': 'am',
?             ^^

+  'period': 'AM',
?             ^^

   'second': 4,
   'timezone': 'UTC',
   'timezone_aware': True,
   'utc_offset': '+0000',
   'week_of_year': 1,
   'weekday': 'Tue',
   'weekday_num': 2,
   'year': 2021}

======================================================================
FAIL: test_date_before_midnight (tests.test_date.MyTests)
Test 'date' 24-hour conversion just before midnight
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/jc/src/jc-1.18.2/tests/test_date.py", line 67, in test_date_before_midnight
    self.assertEqual(jc.parsers.date.parse(self.generic_date_before_midnight, quiet=True), self.generic_date_before_midnight_json)
AssertionError: {'yea[150 chars]d': 'pm', 'timezone': 'UTC', 'utc_offset': '+0[145 chars]True} != {'yea[150 chars]d': 'PM', 'timezone': 'UTC', 'utc_offset': '+0[145 chars]True}
  {'day': 24,
   'day_of_year': 83,
   'epoch': 1616655287,
   'epoch_utc': 1616630087,
   'hour': 11,
   'hour_24': 23,
   'iso': '2021-03-24T23:54:47+00:00',
   'minute': 54,
   'month': 'Mar',
   'month_num': 3,
-  'period': 'pm',
?             ^^

+  'period': 'PM',
?             ^^

   'second': 47,
   'timezone': 'UTC',
   'timezone_aware': True,
   'utc_offset': '+0000',
   'week_of_year': 12,
   'weekday': 'Wed',
   'weekday_num': 3,
   'year': 2021}

======================================================================
FAIL: test_date_pm_ubuntu_20_04 (tests.test_date.MyTests)
Test 'date' on Ubuntu 20.4 with LANG=en_US.UTF-8 (uses 12-hour clock) with PM time
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/tmp/jc/src/jc-1.18.2/tests/test_date.py", line 85, in test_date_pm_ubuntu_20_04
    self.assertEqual(jc.parsers.date.parse(self.ubuntu_20_04_date2, quiet=True), self.ubuntu_20_04_date2_json)
AssertionError: {'yea[149 chars]d': 'pm', 'timezone': 'UTC', 'utc_offset': '+0[145 chars]True} != {'yea[149 chars]d': 'PM', 'timezone': 'UTC', 'utc_offset': '+0[145 chars]True}
  {'day': 23,
   'day_of_year': 82,
   'epoch': 1616557529,
   'epoch_utc': 1616532329,
   'hour': 8,
   'hour_24': 20,
   'iso': '2021-03-23T20:45:29+00:00',
   'minute': 45,
   'month': 'Mar',
   'month_num': 3,
-  'period': 'pm',
?             ^^

+  'period': 'PM',
?             ^^

   'second': 29,
   'timezone': 'UTC',
   'timezone_aware': True,
   'utc_offset': '+0000',
   'week_of_year': 12,
   'weekday': 'Tue',
   'weekday_num': 2,
   'year': 2021}
```
</details>

**Comment by kellyjonbrazil** [comment]
Hi @chocolateboy - thanks for reporting this. This looks like a separate issue around the capitalization of the AM/PM period in the jc output. On my system it looks like AM/PM should be capitalized, though it looks like it is lowercase on your system.

What python version is installed on the system?

**Comment by chocolateboy** [comment]
Thanks for looking into it:

```
$ python --version
3.10.2
```

**Comment by chocolateboy** [comment]
> Maybe I can set the Locale for the tests to C so they don't fail.

Yes, I think that would fix it.

**Comment by kellyjonbrazil** [comment]
Actually, I think I'm just going to `.upper()` the period ('AM'/'PM') text so it's always consistent across systems.

**Comment by chocolateboy** [comment]
Wow, [that](https://github.com/kellyjonbrazil/jc/releases/tag/v1.18.5) was quick. Thank you! :tada:

**Comment by kellyjonbrazil** [comment]
You bet! This should be fixed in v1.18.5. Let me know if you run into any other issues!

**Comment by chocolateboy** [comment]
Will do. Thanks again!
